### ISSUE TITLE
look into a way to reference "this" within a model

### ISSUE THREAD

**Issue by ShMcK** [rematch/rematch#7]
When referencing state within an effect, it would be nice if we could reference state with something like:

```js
reducers: {
  total: 0
},
effects: {
  add: (payload) => this.total + payload
}
```

Is this an idea worth considering?

**Comment by ShMcK** [comment]
See example referenced in #19.

**Comment by ShMcK** [comment]
Some possibilities:

1. Create with objects, but then transform them into classes
2. Use `call` or `bind` to give the model context

**Comment by ShMcK** [comment]
In this case, `this` should actually be `dispatch[modelName]`

**Comment by blairbodnar** [comment]
If in dev mode, after init, run each effect and hook as a string and pattern match for external dispatch calls. Trigger console.warn with a warning message / suggestion if there is a match.

**Comment by ShMcK** [comment]
I currently have an implementation that works in the following scenarios:

1. function scope - PASS
```js
effects: {
  asyncAddOne: async function() {
    await this.addOne()
  }
}
```

2. object function shorthand - PASS
```js
effects: {
  asyncAddOne() {
    this.addOne()
  }
}
```

3. arrow function - FAIL
```js
effects: {
  asyncAddOne: async () => {
    console.log(this) // undefined
    await this.addOne()
  }
}
```

Arrow function will "always" reference its scope, and cannot be `bind`ed to another scope.

**Comment by ShMcK** [comment]
There may be to test if a function is an arrow function, then override it with a regular function.

**Comment by ShMcK** [comment]
It seems the most reliable way to detect an arrow function is by calling `.toString()` on a function, and making a regex. I couldn't find an available regex. But working on one now. Will update this as I go. 

```js
^((async)\s+)?((\(([a-zA-Z0-9_-]+?)?\))|[a-zA-Z0-9_-]+)?(\s?)+=>
```

- [x] () => anything
- [x] (something) => anything
- [ ] (athing, bthing) => something
- [x] async () => something
- [x] async (something) => anything
- [ ] async (athing, bthing) => anything
- [x] ()=>something
- [x] athing=>athing
- [x] athing => athing
- [x] async    () => {}

**Comment by ShMcK** [comment]
Regex to check for arrow functions.
```js
const isArrowFunction = (fn) => {
  const isArrowFunctionRegex = /^((async)\s+)?((\(([a-zA-Z0-9,\s_-]+?)?\))|[a-zA-Z0-9_-]+)?(\s?)+=>/
  return !!fn.toString().match(isArrowFunctionRegex)
}
```

**Comment by ShMcK** [comment]
Unfortunately, this rabbit hole doesn't work. Babel transpiles the arrow function into this mess:

```js
function asyncAddOneArrow() {return _ref2.apply(this, arguments);}
```

**Comment by ShMcK** [comment]
closes in favor of #40. This one is going off topic.
